components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/AbstractHTTPFunction.java

```diff
@@ -73,65 +85,238 @@ public AbstractHTTPFunction() {
         allowedDomains = ConfigProvider.getInstance().getAllowedDomainsForHttpFunctions();
     }
 
-    protected void executeHttpMethod(HttpUriRequest request, Map<String, Object> eventHandlers) {
+    protected void executeHttpMethod(HttpUriRequest clientRequest, Map<String, Object> eventHandlers,
+                                     AuthConfigModel authConfigModel) {
 
         AsyncProcess asyncProcess = new AsyncProcess((context, asyncReturn) -> {
-            JSONObject json = null;
-            int responseCode;
             String outcome;
             String endpointURL = null;
 
-            if (request.getURI() != null) {
-                endpointURL = request.getURI().toString();
+            HttpUriRequest request;
+            try {
+                if (authConfigModel != null) {
+                    AuthConfig authConfig = AuthConfigFactory.getAuthConfig(authConfigModel, context, asyncReturn);
+                    request = authConfig.applyAuth(clientRequest, authConfigModel);
+                } else {
+                    request = clientRequest;
+                }
+
+                if (request.getURI() != null) {
+                    endpointURL = request.getURI().toString();
+                }
+
+                if (!isValidRequestDomain(request.getURI())) {
+                    LOG.error("Request URL does not match with the allowed domain list. Request Url: " +
+                            endpointURL);
+                    asyncReturn.accept(context, Collections.emptyMap(), Constants.OUTCOME_FAIL);
+                } else {
+                    Pair<Integer, Pair<String, JSONObject>> result = executeRequest(request, endpointURL);
+                    if (result.getLeft() >= 400 && result.getLeft() <= 500) {
+                        LOG.error("Error while calling endpoint. Url: " + endpointURL);
+                        result = executeRequestWithRetries(request, endpointURL, requestRetryCount);
```

Comment:
we are not supposed to retry if response status is >= 400 and < 500 right ? Because this means the error is at the http request made by client 
Refer my suggestion on the logic at https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/157#issuecomment-2047019732

Reply:
In that case (https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/157#discussion_r1559232883), we are not retrying at all, right?

`I think we can neglect any redirect response as well from retrying.` -> status is >= 300 and < 400 [Not Retry]
status is >= 200 and < 300 [Not Retry]

If we are not going to retry for `status is >= 400 and < 500` this as well. Then we will retry only in exception scenarios.

Reply:
Add retry boolean to handle this case, instead of staus code

Reply:
Resolved with https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/157/commits/dd0c0830ab854ed1533c10099ec4bb705ddeb218